swag-2024-Q1-openday/slides.md

```diff
@@ -0,0 +1,370 @@
+    - Strong security ensured by built-in AES.
+- Cons:
+    - Not support by native web, require special player.
+    - Require extra bandwidth for SRT machanism maintain stream quality
+
+---
+![original](background.png)
+
+| Protocol             | RTMP        | WebRTC                                         | SRT            |
+|----------------------|-------------|------------------------------------------------|----------------|
+| Supported Codecs     | H.264, AAC  | H.264, VP9, VP8, Opus, G.711 G.722, iLBC, iSAC | Unlimited      |
+| Latency              | < 5s        | < 500ms                                        | < 500ms        |
+| Security             | RTMPS/RTMPE | Built in (DTLS, SRTP)                          | Built in (AES) |
+| Disruption Tolerance | Average     | Good                                           | Good           |
+
+
+---
+[.build-lists: true]
+# [fit] Quiz 1: What video/audio codec does 
+# [fit] **RTMP** support ?
+
+- video: H.264
+- audio: AAC
+
+---
+# [fit] Deep dive into
+# [fit]  **SRT**
+
+---
+[.footer: https://qiita.com/tomoyafujita/items/2e10a9b9d463a36d4a3e]
+![inline](srt-tsbpd.png)
+
+^ 每個packet都會有serial number
+
+---
+![original](background.png)
+
+# Timestamp-Based Packet Delivery
+- Timestamps allow the receiver to **reorder** packets before handover to decoder.
+- Ensuring packets are decoded and displayed in the correct order
+
+--- 
+# Correct order of packets matters
+![inline](srt-decoder-worst-nightmare.png) ![inline](srt-happy-decoder.png)
+
+
+--- 
+[.build-lists: true]
+![original](background.png)
+
+# Automatic Repeat Request (ARQ)
+- Whenever receiver detects packet loss it send NACK to sender to trigger packet retransmission.
+- If packet still in sender's buffer and not determined too late, it will be schedule into queue for sending.
+- Periodically packet loss report.
+
+---
+[.build-lists: true]
+# Packet Delivery Time
+
+$$PktTsbpdTime = TsbpdTimeBase + PktTimestamp + TsbpdDelay + Drift$$
+
+- $$TsbpdTimeBase$$ : Time base reflects the time difference between local clock of the receiver and the sender
+- $$PktTimestamp$$ : Data packet timestamp
+- $$TsbpdDelay$$ : SRT Latency, negotiated between sender and receiver on handshake.
+- $$Drift$$ : Adjust the fluctuations between sender and receiver clock.
+
+---
+
+| Serial Number | Pkt Timestamp | Time Base | SRT Latency | Drift | Packet Delivery Time |
+|---------------|---------------|-----------|-------------|-------|----------------------|
+| 1             | 20            | 40        | 120         | 0     | 180                  |
+| 2             | 40            | 40        | 120         | 0     | 200                  |
+| 5             | 100           | 40        | 120         | 0     | 260                  |
+| 3             | 60            | 40        | 120         | 0     | 220                  |
+| 4             | 80            | 40        | 120         | 0     | 240                  |
```

Comment:
- Delivery time should be first column
- Sorted by delivery time, since that is the timeline to follow along